docs/rfc/rfc-011-delete-gas.md

```diff
@@ -0,0 +1,140 @@
+the mempool to include in the next block. Tendermint uses MaxGas to limit the set of transactions
+it pulls out of the mempool fetching a set of transactions whose sum is less than MaxGas.
+
+By removing gas tracking from Tendermint's mempool, Tendermint will need to provide a way for
+applications to determine an acceptable set of transactions to include in the block.
+
+This is what the new ABCI++ `PrepareProposal` method is useful for. Applications
+that wish to limit the contents of a block by an application-defined limit may
+do so by removing transactions from the proposal it is passed during `PrepareProposal`.
+
+Fully meeting the requirements of the current Gas system would mean adding functionality that is
+not currently provided the `PrepareProposal` method.
+Applications that wish to include _more_ transactions in the proposal than `PrepareProposal`
+initially passed to it cannot do so in the current design. [Issue 7750][issue-7750] has been filed to discuss
+and possibly solve for this.
+
+### Requirement: Handle Transaction Metadata
+
+Moving the gas mechanism into applications adds an additional piece of complexity
+to applications. The application must now track how much gas it expects a transaction
+to consume. The mempool currently handles this bookkeeping responsibility and uses the estimated
+gas to determine the set of transactions to include in the block. In order to task
+the application with keeping track of this metadata, we should make it easier for the
+application to do so. In general, we'll want to keep only one copy of this type
+of metadata in the program at a time, either in the application or in Tendermint.
+
+The following sections are possible solutions to the problem of storing transaction
+metadata without duplication.
+
+#### Metadata Handling: EvictTx Callback
+
+A possible approach to handling transaction metadata is by adding a new `EvictTx`
+ABCI method. Whenever the mempool is removing a transaction, either because it has
+reached its TTL or because it failed `RecheckTx`, `EvictTx` would be called with
+the transaction hash. This would indicate to the application that it could free any
+metadata it was storing about the transaction such as the computed gas fee.
+
+Eviction callbacks are pretty common in caching systems, so this would be very
+well-worn territory.
+
+#### Metadata Handling: Application-Specific Metadata Field(s)
+
+An alternative approach to handling transaction metadata would be would be the
+addition of a new application-metadata field in the `ResponseCheckTx`. This field
+would be a protocol buffer message whose contents were entirely opaque to Tendermint.
+The application would be responsible for marshalling and unmarshalling whatever data
+it stored in this field. During `PrepareProposal`, the application would be passed
+this metadata along with the transaction, allowing the application to use it to perform 
+any necessary filtering.
```

Comment:
I guess I'd like to avoid having to tendermint need to hold on to application data that it's just going to ship back to the application, but if that can be an anti-pattern/non-goal in this work, then I think it's fine

Reply:
> I guess I'd like to avoid having to tendermint need to hold on to application data that it's just going to ship back to the application, but if that can be an anti-pattern/non-goal in this work, then I think it's fine

I agree with this. The main case where I think it might make sense for TM to hold metadata is if there are bits the app wants gossiped along with the txn to other nodes. (If there aren't, IMO, we probably shouldn't have it)